File: scala/sdktest/MessageEntityTest.scala
// Generated basic-flow test for the message entity (model-driven;
// mirrors the java TestEntity generator). A dependency-free scala-cli test
// object driven by SdkEntityTestMain. Runs against the in-memory test
// transport seeded with the shipped MessageTestData.json fixtures.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}

import voxgig.thesmsworkssdk.core.{Helpers, ThesmsworksSDK}
import voxgig.thesmsworkssdk.utility.struct.Struct

object MessageEntityTest {

  def run(rep: SdkTestReport): Unit = {
    rep.scope("message.instance") {
      val testsdk = ThesmsworksSDK.testSDK()
      val ent = testsdk.message(null)
      rep.check("message.instance", ent != null, "expected non-null message entity")
    }

    rep.scope("message.basic") {
      val entityData = Helpers.toMapAny(SdkTestSupport.readJson(
          "../.sdk/test/entity/message/MessageTestData.json"))
      val options = new LinkedHashMap[String, Object]()
      options.put("entity", entityData.get("existing"))
      val client = ThesmsworksSDK.testSDK(options, null)

      val idmap = new LinkedHashMap[String, Object]()
      idmap.put("message01", "MESSAGE01")
      idmap.put("message02", "MESSAGE02")
      idmap.put("message03", "MESSAGE03")
      idmap.put("schedule01", "SCHEDULE01")
      idmap.put("schedule02", "SCHEDULE02")
      idmap.put("schedule03", "SCHEDULE03")
      val now = System.currentTimeMillis()

      // CREATE
      val messageRef01Ent = client.message(null)
      var messageRef01Data = Helpers.toMapAny(Struct.getprop(
          Struct.getpath(entityData, "new.message"), "message_ref01"))
      val messageRef01DataResult = messageRef01Ent.create(messageRef01Data, null)
      messageRef01Data = Helpers.toMapAny(messageRef01DataResult)
      rep.check("message.create.map", messageRef01Data != null, "expected create result to be a map")

      // LOAD
      val messageRef01MatchDt0 = new LinkedHashMap[String, Object]()
      val messageRef01DataDt0Loaded = messageRef01Ent.load(messageRef01MatchDt0, null)
      rep.check("message.load.nonnull", messageRef01DataDt0Loaded != null, "expected load result to be non-null")

      // REMOVE
      val messageRef01MatchRm0 = new LinkedHashMap[String, Object]()
      messageRef01MatchRm0.put("id", messageRef01Data.get("id"))
      messageRef01Ent.remove(messageRef01MatchRm0, null)
    }
  }
}

File: rust/tests/message_entity_test.rs
// Generated basic-flow test for the message entity (model-driven;
// mirrors the go TestEntity generator).

#![allow(unused_variables, unused_mut, unused_imports)]

mod common;

use std::rc::Rc;

use common::*;

use thesmsworks_sdk::core::helpers::{getp, getpath, ja, jo, now_ms, setp, to_map};
use thesmsworks_sdk::utility::voxgigstruct as vs;
use thesmsworks_sdk::{test_sdk, Entity, ThesmsworksEntity, ThesmsworksSDK, Value};

#[test]
fn message_entity_instance() {
    let testsdk = test_sdk(Value::Noval, Value::Noval);
    let ent = testsdk.message(Value::Noval);
    assert_eq!(ent.get_name(), "message");
}

#[test]
fn message_entity_basic() {
    let setup = message_basic_setup(Value::Noval);
    // Per-op sdk-test-control.json skip — the basic test exercises a flow
    // with multiple ops; skipping any op skips the whole flow.
    let mode = if setup.live { "live" } else { "unit" };
    for op in ["create", "load", "remove"] {
        let (skip, reason) = is_control_skipped("entityOp", &format!("message.{}", op), mode);
        if skip {
            let reason = if reason.is_empty() {
                "skipped via sdk-test-control.json".to_string()
            } else {
                reason
            };
            eprintln!("skip: {}", reason);
            return;
        }
    }
    // The basic flow consumes synthetic IDs from the fixture. In live mode
    // without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only {
        eprintln!("skip: live entity test uses synthetic IDs from fixture — set THESMSWORKS_TEST_MESSAGE_ENTID JSON to run live");
        return;
    }
    let client = setup.client.clone();
    // CREATE
    let message_ref01_ent = client.message(Value::Noval);
    let message_ref01_data = to_map(&getp(
        &getpath(&["new", "message"], &setup.data),
        "message_ref01",
    ));

    let message_ref01_data_result = message_ref01_ent
        .create(message_ref01_data.clone(), Value::Noval)
        .expect("create failed");
    let message_ref01_data = to_map(&message_ref01_data_result.data(None));
    assert!(
        matches!(message_ref01_data, Value::Map(_)),
        "expected create result to be a map"
    );
    assert!(
        !getp(&message_ref01_data, "id").is_noval(),
        "expected created entity to have an id"
    );

    // LOAD
    let message_ref01_match_dt0 = jo(vec![("id", getp(&message_ref01_data, "id"))]);
    let message_ref01_data_dt0_loaded = message_ref01_ent
        .load(message_ref01_match_dt0.clone(), Value::Noval)
        .expect("load failed");
    let message_ref01_data_dt0_load_result = to_map(&message_ref01_data_dt0_loaded.data(None));
    assert!(
        matches!(message_ref01_data_dt0_load_result, Value::Map(_)),
        "expected load result to be a map"
    );
    assert_eq!(
        getp(&message_ref01_data_dt0_load_result, "id"),
        getp(&message_ref01_data, "id"),
        "expected load result id to match"
    );

    // REMOVE
    let message_ref01_match_rm0 = jo(vec![("id", getp(&message_ref01_data, "id"))]);
    message_ref01_ent
        .remove(message_ref01_match_rm0.clone(), Value::Noval)
        .expect("remove failed");

}

fn message_basic_setup(extra: Value) -> EntityTestSetup {
    load_env_local();

    let mut entity_data_file = manifest_dir();
    entity_data_file.push("..");
    entity_data_file.push(".sdk");
    entity_data_file.push("test");
    entity_data_file.push("entity");
    entity_data_file.push("message");
    entity_data_file.push("MessageTestData.json");

    let entity_data = read_json(&entity_data_file);

    let options = jo(vec![("entity", getp(&entity_data, "existing"))]);

    let client = test_sdk(options, extra.clone());

    // Generate idmap via transform, matching the TS pattern.
    let idmap = vs::transform(
        &ja(vec![Value::str("message01"), Value::str("message02"), Value::str("message03"), Value::str("schedule01"), Value::str("schedule02"), Value::str("schedule03")]),
        &jo(vec![(
            "`$PACK`",
            ja(vec![
                Value::str(""),
                jo(vec![
                    ("`$KEY`", Value::str("`$COPY`")),
                    (
                        "`$VAL`",
                        ja(vec![
                            Value::str("`$FORMAT`"),
                            Value::str("upper"),
                            Value::str("`$COPY`"),
                        ]),
                    ),
                ]),
            ]),
        )]),
        None,
    )
    .unwrap_or_else(|_| Value::empty_map());

    // Detect ENTID env override before env_override consumes it. When live
    // mode is on without a real override, the basic test runs against
    // synthetic IDs from the fixture and 4xx's.
    let entid_env_raw = std::env::var("THESMSWORKS_TEST_MESSAGE_ENTID").unwrap_or_default();
    let idmap_overridden =
        !entid_env_raw.trim().is_empty() && entid_env_raw.trim().starts_with('{');

    let env = env_override(jo(vec![
        ("THESMSWORKS_TEST_MESSAGE_ENTID", idmap.clone()),
        ("THESMSWORKS_TEST_LIVE", Value::str("FALSE")),
        ("THESMSWORKS_TEST_EXPLAIN", Value::str("FALSE")),
        ("THESMSWORKS_APIKEY", Value::str("")),
    ]));

    let idmap_resolved = match to_map(&getp(&env, "THESMSWORKS_TEST_MESSAGE_ENTID")) {
        Value::Map(m) => Value::Map(m),
        _ => to_map(&idmap),
    };

    let live = getp(&env, "THESMSWORKS_TEST_LIVE") == Value::str("TRUE");

    let client = if live {
        let merged = vs::merge(
            // live_client_options() FIRST, so the generated entries below win:
            // sdk-test-control.json's test.client.options adds to the live
            // client, it does not redirect it.
            &ja(vec![
                live_client_options(),
                jo(vec![("apikey", getp(&env, "THESMSWORKS_APIKEY"))]),
                // A NON-NODE later entry REPLACES the accumulated map in
                // vs::merge, and the normal call passes Value::Noval - so a
                // a bare extra discarded live_client_options() and the
                // apikey/server map above it, and the live client was
                // constructed with nothing.
                match extra {
                    Value::Map(m) => Value::Map(m),
                    _ => Value::empty_map(),
                },
            ]),
            None,
        );
        ThesmsworksSDK::new(to_map(&merged))
    } else {
        client
    };

    EntityTestSetup {
        client,
        data: entity_data,
        idmap: idmap_resolved,
        env: env.clone(),
        explain: getp(&env, "THESMSWORKS_TEST_EXPLAIN") == Value::str("TRUE"),
        live,
        synthetic_only: live && !idmap_overridden,
        now: now_ms(),
    }
}
